taf/tests/test_updater/test_update_library/test_validation_and_sync.py

```diff
@@ -0,0 +1,214 @@
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(update_expiration_dates)
+    setup_manager.add_task(add_valid_target_commits)
+    setup_manager.add_task(
+        revert_last_validated_commit, kwargs={"client_dir": client_dir}
+    )
+    setup_manager.execute_tasks()
+
+    assert_repositories_updated(client_dir, origin_auth_repo)
+
+
+@pytest.mark.parametrize(
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_target_repo_not_in_sync_invalid_commits(origin_auth_repo, client_dir):
+    """
+    Same as 2, but target repositories contain invalid commits following commits which were manually pulled using git pull.
+    Expect all repositories to be updated up to the last validated commits and invalid commits to be removed
+    """
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(add_valid_unauthenticated_commits)
+    setup_manager.add_task(update_expiration_dates)
+    setup_manager.add_task(
+        sync_target_repos_with_remote,
+        kwargs={"origin_auth_repo": origin_auth_repo, "client_dir": client_dir},
+    )
+    setup_manager.add_task(
+        revert_last_validated_commit, kwargs={"client_dir": client_dir}
+    )
+    setup_manager.execute_tasks()
+
+    assert_repositories_updated(client_dir, origin_auth_repo)
+
+
+@pytest.mark.parametrize(
+    "origin_auth_repo",
+    [
+        {
+            "targets_config": [
+                {"name": "target1", "allow_unauthenticated_commits": True},
+                {"name": "target2", "allow_unauthenticated_commits": True},
+            ],
+        },
+    ],
+    indirect=True,
+)
+def test_mixed_target_repo_states(origin_auth_repo, client_dir):
+    """
+    Client's auth repo is not on last validated commit, but target commits are all over the place.
+    Some are after last validated, some are before.
+    At the moment, i believe that the validation will start from the beginning in such a case and will be successful
+    if there are no invalid commits.
+    """
+    client_target_repo_path = (
+        client_dir
+        / origin_auth_repo.name
+        / "targets/test_remove_commits_from_target_repo0/target1"
+    )
+
+    remove_commits(str(client_target_repo_path))
+    clone_repositories(
+        origin_auth_repo,
+        client_dir,
+    )
+    setup_manager = SetupManager(origin_auth_repo)
+    setup_manager.add_task(add_valid_target_commits)
+    setup_manager.add_task(update_expiration_dates)
+    setup_manager.add_task(
+        remove_commits, kwargs={"repo_path": client_target_repo_path, "num_commits": 1}
```

Comment:
We want the target repos to be completely inconsistent. E.g target 1 is behind last validated commit, target 2 was updated manually